core/src/main/kotlin/xtdb/ICursor.kt

```diff
@@ -77,7 +81,65 @@ interface ICursor : Spliterator<RelationReader>, AutoCloseable {
             override fun getComparator(): Comparator<in RelationReader>? = inner.comparator
         }
 
+        private class TracingCursor(
+            private val inner: ICursor,
+            private val tracer: Tracer,
+            private val parentSpan: Span,
+            private val clock: InstantSource = InstantSource.system()
+        ) : ICursor by inner {
+            private var pageCount: Int = 0;
+            private var rowCount: Long = 0;
+            private var started = false
+            private var startTime = clock.instant()
+            private var timeToFirstPage: Duration? = null
+            private var totalTime: Duration = Duration.ZERO
+            private var span: Span? = null
+
+            override fun tryAdvance(c: Consumer<in RelationReader>): Boolean {
+                if (!started) {
+                    started = true
+                    startTime = clock.instant()
+                    span = tracer.nextSpan(parentSpan)!!
+                        .name("query.cursor.${inner.cursorType}")
+                        .tag("cursor.type", inner.cursorType)
+                        .start()
+                }
+                val pageStart = clock.instant()
+
+                return inner.tryAdvance { rel ->
+                    val pageTime = Duration.between(pageStart, clock.instant())
+                    timeToFirstPage = timeToFirstPage ?: pageTime
+                    rowCount += rel.rowCount
+                    pageCount++
+                    c.accept(rel)
+                }.also {
+                    totalTime += Duration.between(pageStart, clock.instant())
+                }
+            }
+
+            override fun close() {
+                span!!.tag("cursor.total_time_ms", totalTime.toMillis().toString())
+                timeToFirstPage?.let { ttfp ->
+                    span!!.tag("cursor.time_to_first_page_ms", ttfp.toMillis().toString())
+                }
+                span!!.tag("cursor.page_count", pageCount.toString())
+                span!!.tag("cursor.row_count", rowCount.toString())
+                inner.close()
+                val endTime = startTime.plus(totalTime)
```

Comment:
I might be tempted to keep endTime as a var on the class too, given that you can manually supply it here - i.e. bump it at the end of each call to tryAdvance, then use that time here?